ft_prepare_sourcemodel.m

```diff
@@ -642,6 +646,37 @@
       % copy the boolean fields over
       sourcemodel = copyfields(mnigrid, sourcemodel, booleanfields(mnigrid));
     end
+
+  case 'basedoncentroids'
+    %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+    % compute the centroids of each element of a volumetric mesh and use
+    % them as sources. assume all inside.
+    %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+    % this is a model with hexaheders or tetraheders
+    if isfield(headmodel, 'tet')
+      % the subsequent code works both for tetraheders or hexaheders, but assumes the volume elements to be called "hex"
+      headmodel.hex = headmodel.tet;
+      headmodel = rmfield(headmodel, 'tet');
+    end
+
+    % determine the size of the relevant elements
+    numhex = length(headmodel.hex);
+    numelem = length(headmodel.hex(1,:));
+
+    sourcemodel.pos = zeros(numhex,3);
+
+    for i=1:length(headmodel.hex)
+        matr_help = zeros(numelem,3);
+        for j=1:numelem
+            matr_help(j,:) = headmodel.pos(headmodel.hex(i,j),:);
+        end
+        sourcemodel.pos(i,:) = mean(matr_help);
+    end
+
+    sourcemodel.tissue = headmodel.tissue;
+    sourcemodel.tissuelabel = headmodel.tissuelabel;
+    sourcemodel.unit    = cfg.unit;
+    sourcemodel.inside  = true(size(sourcemodel.pos,1),1);
```

Comment:
Once you add the inside field, it will (later) be assumed to be correct and hence not revised any more. If you don't know what should be inside and outside yet, better don't add it: the code later will deal with it (and potentially set all positions to inside)

Reply:
I modified it. 
with a hexa mesh with 5 compartments (scalp, skull, csf, brain, lesion) the inside sources are in brain+csf.